### ISSUE TITLE
incorrect requirements for pseudo header field mapping for asterisk form

### ISSUE THREAD

**Issue by reschke** [httpwg/http2-spec#828]
I'm struggling with <https://greenbytes.de/tech/webdav/rfc7540.html#HttpRequest>:

> The :authority pseudo-header field includes the authority portion of the target URI ([RFC3986], Section 3.2). The authority MUST NOT include the deprecated userinfo subcomponent for http or https schemed URIs.
>
> To ensure that the HTTP/1.1 request line can be reproduced accurately, this pseudo-header field MUST be omitted when translating from an HTTP/1.1 request that has a request target in origin or asterisk form (see [RFC7230], Section 5.3).

If I have

~~~
  OPTIONS * HTTP/1.1
  Host: example.com
~~~

and wanted to map this to H2, why would I omit ":authority" here?

Wild guess: the spec should say "origin or authority form", not "origin or asterisk form"?

**Comment by Lukasa** [comment]
It definitely shouldn't say "origin or authority form". Authority form is well-covered by the spec: the `:path` pseudo-header absolutely may not contain the authority component of the target URI, so if it isn't put into the `:authority` pseudo-header it has nowhere to go and would be lost.

For asterisk form though I'm as much guessing as you are, but I can provide two alternative theories. The first is predicated on the use of the phrase "target URI". An OPTIONS request with a request target of `*` does not really _have_ a target URI for the `:authority` pseudo-header to carry the authority portion _of_. However, that's not terribly compelling, given that -semantics fairly heavily implies that `:authority` should just have whatever is in `Host` (which it cannot do for CONNECT requests and so we probably need to reword that text in -semantics?).

The _better_ argument is that `OPTIONS *` doesn't meaningfully act on a host at all. From -semantics 9.3.7:

> An OPTIONS request with an asterisk ("*") as the request target (Section 7.1) applies to the server in general rather than to a specific resource.

The word here is "server", not "origin". It isn't clear that the -semantics specification envisages the possibility that a server may care what the value of the `Host` header is when responding to this request.

With that said, I do think we need to clarify, perhaps with the WG, whether the Host: header is expected to be meaningful for `OPTIONS *`. If it is, we clearly need to either allow it in the `:authority` pseudo-header or provide guidance that you should put it in `Host` (for some weird reason).

**Comment by reschke** [comment]
"OPTIONS *" indeed is weird, and it would be great if we could get rid of it. But we can't.

The core spec does not state that "OPTIONS *" can not vary on "Host", so there may be cases where it does. So I believe H2 needs to describe the mapping for that case.

**Comment by Lukasa** [comment]
I think if we're happy with the idea that `OPTIONS *` can vary on `Host` then the obvious change is to just delete "or asterisk" from the text cited in the original post.

**Comment by reschke** [comment]
Awesome :-)

**Comment by martinthomson** [comment]
Yes, I think that would be the best outcome.  Using `:authority` is consistent with other h2 usage.

Though as @mnot noted privately, the fact that this question has not arisen might be evidence that `OPTIONS *` is less widely implemented than we thought.

**Comment by Lukasa** [comment]
Hang on hang on hang on. When I was writing the patch I went to double-check the citations and got very confused. The current text reads:

> To ensure that the HTTP/1.1 request line can be reproduced accurately, this pseudo-header field MUST be omitted when translating from an HTTP/1.1 request that has a request target in origin or asterisk form (see [RFC7230], Section 5.3).

Reproducing from the current -messaging draft, here are the definitions of the request target forms:

> The most common form of request-target is the origin-form.
>
>     origin-form    = absolute-path [ "?" query ]

> When making a request to a proxy, other than a CONNECT or server-wide OPTIONS request (as detailed below), a client must send the target URI in absolute-form as the request-target.
>
>     absolute-form  = absolute-URI

> The authority-form of request-target is only used for CONNECT requests (Section 9.3.6 of [Semantics]).
>
>     authority-form = authority

> The asterisk-form of request-target is only used for a server-wide OPTIONS request (Section 9.3.7 of [Semantics]).
>
>    asterisk-form  = "*"

The original draft says that you should only produce an `:authority` pseudo-header field _when forwarding a HTTP/1.1 request_ in the cases where the _request target_ normally had an authority within it, either in the absolute-form case or the authority-form case.

I think then the original text is right, just confusing, and made more confusing by the fact that it's followed by a line talking about clients that generate HTTP/2 requests directly that makes no sense.

What the original text is trying to say, I think, is that in HTTP/1.1 the maximum number of authorities that may be present in control data is two: one in the request target and one in the (mandatory) Host header. This can only happen for requests which use absolute-form or authority-form: for requests that use origin-form or asterisk-form they will only have had one in the Host header.

The implication here then is that there is nothing wrong with sending `:authority` for an `OPTIONS *` request, but only if you started out speaking HTTP/2. If you were forwarding a HTTP/1.1 request then you should leave the authority in the Host component. This is intended to make it easier to reconstruct the request at the end: you don't have to look at the `:method` and `:path` to decode it appropriately.

What makes this awkward is that the defensiveness is partially unnecessary. -messaging says:

> A client MUST send a Host header field in all HTTP/1.1 request messages. If the target URI includes an authority component, then a client MUST send a field value for Host that is identical to that authority component, excluding any userinfo subcomponent and its "@" delimiter (Section 4.2.1 of [Semantics]).

This means that in the case of `authority-form` the `:authority` and `Host` headers MUST be identical (the `authority-form` also forbids the userinfo subcomponent). The restriction doesn't really make sense here: we could have absolutely collapsed this into `:authority`.

Unfortunately, -messaging does still leave an out in the case of `absolute-form` request targets, where everyone is required to ignore the Host header field. This probably justifies leaving that case at the very least.

To sum up then, I think the text is right, but may want clarification.

**Comment by reschke** [comment]
Hmmm.

We have 1.:

> To ensure that the HTTP/1.1 request line can be reproduced accurately, this pseudo-header field MUST be omitted when translating from an HTTP/1.1 request that has a request target in origin or asterisk form (see [RFC7230], Section 5.3). 

and 2.:

> All HTTP/2 requests MUST include exactly one valid value for the :method, :scheme, and :path pseudo-header fields, unless it is a CONNECT request (Section 8.3).

Aren't these two in conflict for "OPTIONS *"?

**Comment by martinthomson** [comment]
So as @Lukasa says, the important qualification here is that this applies only to translation from HTTP/1.1.  I missed that important point.  If you are newly creating an HTTP/2 request then you can use :authority.

What is awkward here is that we don't really say what to do about :scheme for asterisk form requests.  It would be reasonable to assume that it takes on the default value (https) from a translated request and that it is also set to the same (https) when the h2 request is created directly.  But it doesn't really *say* that directly.  Clarification on this point would be welcome.

**Comment by Lukasa** [comment]
@reschke I think 2. is flat out wrong and should be amended, it is not in line with the rest of the spec.

**Comment by Lukasa** [comment]
Nope, it isn't wrong, it's fine. 1. discusses `:authority`, 2. discusses `:method`, `:path`, and `:scheme`.

**Comment by Lukasa** [comment]
> What is awkward here is that we don't really say what to do about :scheme for asterisk form requests. It would be reasonable to assume that it takes on the default value (https) from a translated request and that it is also set to the same (https) when the h2 request is created directly. But it doesn't really say that directly. Clarification on this point would be welcome.

After trying to construct wording for this I came to the conclusion that this is harder than it might seem. What actually _is_ the scheme for a request in asterisk form? Not in HTTP/2, but what answer would we give if we were writing this for -semantics? I think this is the lens through which I want to word this in the spec, and I don't know that I have a good answer.

**Comment by martinthomson** [comment]
I was going to say that for `:scheme` we open an issue on -semantics, but conclude that the scheme was `https`.  Nothing else really makes sense.  

`CONNECT` is funny in a deliberate fashion (as Erik Nygren points out, the target of a `CONNECT` isn't a resource that is controlled by the thing that serves the request, which makes for all sorts of awkward).  So maybe it gets a pass for being old and strange.

On the other hand, I don't see why `OPTIONS *` needs to be exempt from this; it's just a request for an HTTP resource that can't be identified with a URI because it doesn't have a path component (or query, or fragment).

**Comment by mnot** [comment]
Just popping up a level -- h2bis should describe whatever it needs to here in terms of Semantics, not Messaging. If it needs to refer to 1.1 specifically, something's broken.
